controllers/front/OrderConfirmationController.php

```diff
@@ -45,31 +48,92 @@ public function init()
     {
         parent::init();
 
+        // If are coming to this page to finish free order, we need to run additional logic
         if (true === (bool) Tools::getValue('free_order')) {
             $this->checkFreeOrder();
         }
 
-        $this->id_cart = (int) (Tools::getValue('id_cart', 0));
-
-        $redirectLink = 'index.php?controller=history';
-
+        // We get the required data from URL
         $this->id_module = (int) (Tools::getValue('id_module', 0));
-        $this->id_order = Order::getIdByCartId((int) ($this->id_cart));
+        $this->id_order = (int) (Tools::getValue('id_order', 0));
         $this->secure_key = Tools::getValue('key', false);
-        $order = new Order((int) ($this->id_order));
+        $this->order = new Order((int) ($this->id_order));
```

Comment:
I think this change is important. `$order` ⏩  `$this->order` can you explain it ?

Reply:
@matks It's because previously, the Order was created twice for nothing - `$order = new Order(Order::getIdByCartId((int) ($this->id_cart)));`. Now I initialize the object only once in init() and save it to the attribute, so we can use it in all other methods.